### DISCUSSION TITLE
Does autocomplete/intellisense only work with TS?

### DISCUSSION THREAD

**Discussion by shainegordon** [vuejs/pinia#531] [Help and Questions] [answered]
Just trying to understand when and where autocomplete/intellisense would work in Pinia, or if this is a config issue on my side.

Using WebStorm and Nuxt

In a store:
no `dog` appearing in autocomplete
![image](https://user-images.githubusercontent.com/1714867/121379985-d1d09900-c944-11eb-8c1b-e8f5dae34d15.png)

calling another action does work (as expected)
![image](https://user-images.githubusercontent.com/1714867/121380442-2a079b00-c945-11eb-8652-0b5a7c23bd7d.png)

in a component:
unknown action `fetchUser` and state `user`
![image](https://user-images.githubusercontent.com/1714867/121380738-6cc97300-c945-11eb-88cf-51a547203d3e.png)

Is this a JS-only restriction, and the above would work correctly with TS?

All in all though, really like this library, excellent work :+1:

**Comment by posva** [comment]
> All in all though, really like this library, excellent work 👍

Thanks 😃 

It should work with JS too but note your `user` is initialized to `null`. In the docs there is an example about typing even in JS with comments. I don't know about the problem with `fetchUser()`. Maybe having a `jsconfig.json` (like a tsconfig.json) helps the IDE and TS  to make it work.

**Comment by ttkoma** [comment]
I have the same problem

**Comment by BenShelton** [comment, marked as answer]
See documentation for adding TS intellisense in WebStorm:
- https://www.jetbrains.com/help/webstorm/auto-completing-code.html#ws_completion_from_tw_language_service
- https://www.jetbrains.com/help/webstorm/typescript-support.html#ws_ts_use_ts_service_checkbox

Although I don't have this IDE to try, creating a `jsconfig.json` with `"allowJs": true` set should be enough. Example `jsconfig.json`:

```json
{
  "include": [
    "./src/**/*"
  ],
  "compilerOptions": {
    "allowJs": true
  }
}
```

Make sure to restart the service after these changes:
> Click the TypeScript widget on the Status bar and select Restart TypeScript Service from the list.

**Comment by mousesama** [comment]
I seem to be having a very similar issue where even going by the examples in docs, I can access my store properties, getters, and actions just fine in the script. I return the whole store from setup() to use in my templates, which works except WebStorm Intellisense does not detect any of my custom properties and functions. When I use the properties, getters, or actions in the template, they work as one would expect but WebStorm just can't seem to find them even though it sees the other built-in properties of the store (see pic). Adding `allowJS`  to `tsconfig.json` as described above unfortunately does not help.

![peMtA85YFM](https://user-images.githubusercontent.com/68248807/140203938-f34b8c54-0503-4ea1-8af9-2893a3b02c47.png)

This is my simple test store:
```ts
export const useResources = defineStore('resources', {
    state: (): Resources => ({
        name: "name",
        money: 100
    }),
    getters: {
        getMoney(state) {
            return state.money
        }
    },
    actions: {
        addMoney() {
            this.money++
        },
        addMoreMoney(value: number) {
            this.money += value
        }
    }
})
```
which I then use in my Vue component like so:
```ts
<script lang="ts">
import { useResources } from "@/stores/resources";
import { defineComponent } from "vue";

export default defineComponent({
  name: "TheResources",
  setup() {
    const resourcesStore = useResources()

    return {
      resourcesStore,
      increment: resourcesStore.addMoney
    }
  }
})
</script>
```

**Comment by mousesama** [comment]
I managed to find the issue and a workaround for it, more details in this opened issue: https://github.com/posva/pinia/issues/772

**Comment by joshkrz** [comment]
I had to do a couple of things to get autocompletion working. I'm using PHPStorm 2021.2.3 (WebStorm under the hood), it is also worth noting I'm _not_ using Typescript in my projects.

- Move the `id` parameter in the Pinia store into the options object.
  - https://github.com/posva/pinia/discussions/774#discussioncomment-1597802
- Export the defineStore function by default and import to components without destructuring.

Adding a `jsconfig.json` seemed to have no effect for me.

EDIT: Just realised that autocompletion inside of the store in actions and getters does not work, even with the `jsconfig.json`. I'll post back if I find a solution.

![SCR-20211125-bvt](https://user-images.githubusercontent.com/6578760/143406836-7cf98995-4aa3-4723-beef-a4645ec362f2.png)

![SCR-20211125-c28](https://user-images.githubusercontent.com/6578760/143407942-c73550c5-e3fd-4efb-88d3-3a6ec20adc0b.png)

**Reply by Meller008** [reply]
Yes it worked for me. But I don't like the default export.

**Reply by zjssos** [reply]
maybe you can do it like 
menus.js
`export default defineStore({
  id: 'menus',
 ...
})
`
index.js
`
import menus from '@/store/modules/menus'
export const menusStore = menus
`

**Comment by keyding** [comment]
This is OK.

![image](https://user-images.githubusercontent.com/20431351/146291570-f66ce460-d6a0-4492-af5f-425497312d59.png)

**Comment by djpeach** [comment]
# Looks to be the default import. 

## This fails 
I created a store with the id param, some state and an action and exported it as a const:
<img width="980" alt="image" src="https://user-images.githubusercontent.com/31779571/163650087-cf90ab48-c06e-4165-b864-1c8f4fa9b5e8.png">

I then import it and the type of the store loses the actions and state:
<img width="1009" alt="image" src="https://user-images.githubusercontent.com/31779571/163650104-4f600398-240f-4420-93d6-c1a2452bd923.png">

## This succeeds
However, exporting it as the default export and then importing it keeps the state and actions in the type, and the resolution works:
<img width="1011" alt="image" src="https://user-images.githubusercontent.com/31779571/163650146-fe46232a-d668-4ae1-9b9e-0e2563412440.png">
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/31779571/163650162-e213a2c4-0063-4099-8649-dabe451e24fd.png">

## Note
So looks like the answer for now is use a default export. If you want to have a name in the store file as well like I do, you can still name the const and export it as the default: 
<img width="1015" alt="image" src="https://user-images.githubusercontent.com/31779571/163650207-56374564-489c-48a1-8ed3-8ae8a451f4e3.png">

**Reply by taisha** [reply]
doesn't  work with webstorm 2022.1

**Reply by m4heshd** [reply]
Unfortunately it only works with actions. Not with state or getters. _(2022.1)_

**Comment by me-v2** [comment]
how fix it ?
